internal/gatewayapi/resource/strategy.go

```diff
@@ -0,0 +1,197 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package resource
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	apiextensionshelpers "k8s.io/apiextensions-apiserver/pkg/apihelpers"
+	apiextensionsinternal "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	structuraldefaulting "k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting"
+	apiservervalidation "k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+	"k8s.io/apiextensions-apiserver/pkg/crdserverscheme"
+	"k8s.io/apiextensions-apiserver/pkg/registry/customresource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+// versionedCreateStrategyMap only stores latest version of CreateStrategy.
+type versionedCreateStrategyMap map[string]rest.RESTCreateStrategy
+
+func (v versionedCreateStrategyMap) Validate(kind string, obj runtime.Object) error {
+	createStrategy, ok := v[kind]
+	if !ok {
+		return nil
+	}
+
+	errList := createStrategy.Validate(context.TODO(), obj)
+	if errList != nil {
+		return errList.ToAggregate()
+	}
+
+	return nil
+}
+
+var (
+	// crdCreateStrategyMap stores CreateStrategy for supported CRDs.
+	// Since it's a lazy loading map, calling getCreateStrategyMapForCRDs
+	// before using this map, directly using this map is not recommended.
+	crdCreateStrategyMap versionedCreateStrategyMap
+	once                 sync.Once
+)
+
+func getCreateStrategyMapForCRDs() versionedCreateStrategyMap {
+	once.Do(func() {
+		// This error should never happen. If it does, stop loading strategy map.
+		crds, _ := loadCRDs()
+		if len(crds) == 0 {
+			return
+		}
+		crdCreateStrategyMap = make(versionedCreateStrategyMap, len(crds))
+		for _, crd := range crds {
+			vcs, _ := getVersionedCreateStrategyForCRD(crd)
+			crdCreateStrategyMap[crd.Spec.Names.Kind] = vcs
+		}
+	})
+	return crdCreateStrategyMap
+}
+
+// getVersionedCreateStrategyForCRD will only get latest version of the input CRD and
```

Comment:
is this what `kubectl validate` uses under the hood ?

Reply:
sort of, we both use Strategy to validate resources before creating.

Reply:
yeah would make it more maintainable imo